torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -32,14 +34,13 @@ void RpcAgent::start() {
   startImpl();
 }
 
-void RpcAgent::cleanup() {
+void RpcAgent::shutdown() {
   rpcAgentRunning_.store(false);
```

Comment:
I think there's actually a race condition here - you might need to temporarily grab rpcRetryMutex_ when setting for this var for the exit to work properly.

(I just noticed a similar issue in a ParallelNative.cpp PR yesterday)

It's non-intuitive. but consider the following:
```
  - retryThread:       if (!rpcAgentRunning_.load())    // still true, so we do nothing
  - rpcAgent:     rpcAgentRunning_.store(false);	  // no it's false
                             rpcRetryMapCV_.notify_one();	  // no waiters in the cv, so it does nothing
   - retryThread:       rpcRetryMapCV_.wait_until(lock, earliestTimeout);   // shoot, we basically sleep forever if it's empty    

Where the mutex adds value is  in  protecting against mutation in the critical section between the waiter noticing one value and going to sleep.

This seems to be somewhat related to the purpose of your change here, do you want to fix along with this?

Reply:
I see, so in the above case, `shutdown` will only grab rpcRetryMutex_ when the retry thread unlocks the lock upon entering the `wait_until` - so the var can't be set to False between checking the value and the retry thread sleeping. Makes sense, I'll fix this.